PMNS_ScalarNSI.cxx

```diff
@@ -0,0 +1,410 @@
+  }
+
+  return fEta[flvi][flvj];
+
+}
+/// Set lightest neutrino mass
+void PMNS_ScalarNSI::SetM(double m){
+
+fM=m;
+
+}
+//......................................................................
+///
+/// Set eta_ee parameter
+///
+/// This will check if value is changing to keep track of whether
+/// the eigensystem needs to be recomputed.
+///
+/// @param a - The value of the real parameter eta_ee
+///
+void PMNS_ScalarNSI::SetEta_ee(double a){ SetEta(0,0, a, 0); }
+
+//......................................................................
+///
+/// Set eta_mumu parameter
+///
+/// This will check if value is changing to keep track of whether
+/// the eigensystem needs to be recomputed.
+///
+/// @param a - The value of the real parameter eta_mumu
+///
+void PMNS_ScalarNSI::SetEta_mumu(double a){ SetEta(1,1, a, 0); }
+
+//......................................................................
+///
+/// Set eta_tautau parameter
+///
+/// This will check if value is changing to keep track of whether
+/// the eigensystem needs to be recomputed.
+///
+/// @param a - The value of the real parameter eta_tautau
+///
+void PMNS_ScalarNSI::SetEta_tautau(double a){ SetEta(2,2, a, 0); }
+
+//......................................................................
+
+void PMNS_ScalarNSI::SetEta_emu(double a, double phi){ SetEta(0,1, a, phi); }
+
+//......................................................................
+///
+/// Set eta_mumu parameter
+///
+/// This will check if value is changing to keep track of whether
+/// the eigensystem needs to be recomputed.
+///
+/// @param a - The value of the real parameter eta_mumu
+///
+void PMNS_ScalarNSI::SetEta_mutau(double a, double phi){ SetEta(1,2, a, phi); }
+
+//......................................................................
+///
+/// Set eta_tautau parameter
+///
+/// This will check if value is changing to keep track of whether
+/// the eigensystem needs to be recomputed.
+///
+/// @param a - The value of the real parameter eta_tautau
+///
+void PMNS_ScalarNSI::SetEta_etau(double a, double phi){ SetEta(0,2, a, phi); }
+
+
+////Get lightest neutrino mass
+
+
+double PMNS_ScalarNSI::GetM(){ return fM;}
+/// Set eps_mutau parameter
+///
+/// This will check if value is changing to keep track of whether
+/// the eigensystem needs to be recomputed.
+///
+/// @param a   - The absolute value of the parameter eps_mutau
+///
+/// Build the full Hamiltonian in matter.
+///
+/// Here we divide the mass squared matrix Hms by the 2E
+/// to obtain the vacuum Hamiltonian in eV. Then, the matter
+/// potential is added to the electron component.
+///
+void PMNS_ScalarNSI::UpdateHam()
+{
+
+  double lv = 2 * kGeV2eV*fEnergy;     // 2E in eV 
+
+  double kr2GNe = kK2*M_SQRT2*kGf;
+  kr2GNe *= fPath.density * fPath.zoa; // Matter potential in eV
+
+  // PMNS matrix U
+  
+  complexD U[3][3];
+  int i,j,k;
+  for( i=0;i<3;i++)
+  {
+    
+     for( j=0;j<3;j++)
+     {
+        U[i][j]=fEvec[i][j];
+     }
+  }
```

Comment:
Why copy the fEvec into U? You can just use fEvec directly.

However, I'm not sure here you're doing what you want. fEvec is a general PMNS matrix in matter. You probably need to make sure it is initialized as a vacuum special case somewhere.